src/samples/audio/smart_amp_test.c

```diff
@@ -32,20 +36,65 @@ struct smart_amp_data {
 	struct comp_buffer *feedback_buf; /**< feedback source buffer */
 	struct comp_buffer *sink_buf; /**< sink buffer */
 
+#ifndef __ZEPHYR__
+	struct k_spinlock lock; /**< protect feedback_buf updated */
+	k_spinlock_key_t key;
+#else
+	struct k_mutex lock;
+#endif
+
 	smart_amp_proc process;
 
 	uint32_t in_channels;
 	uint32_t out_channels;
 };
 
+#ifdef __ZEPHYR__
+
+static void sa_lock(struct smart_amp_data *sad)
+{
+	k_mutex_lock(&sad->lock, K_FOREVER);
+}
+
+static void sa_unlock(struct smart_amp_data *sad)
+{
+	k_mutex_unlock(&sad->lock);
+}
+
+static void sa_lock_init(struct smart_amp_data *sad)
+{
+	k_mutex_init(&sad->lock);
+}
```

Comment:
Can we revisit this in a subsequent PR (as there are others doing the same) and use k_mutex_ API directly (where we will define this for xtos as a spinlock).

Reply:
sure, thanks

Reply:
ok, since we are updating the PR due to other issues, let fix this one too. Please add the k_mutex() calls to xtos/include/rtos/spinlock.h (and map them to use spinlocks).

Reply:
@lgirdwood are we sure that this is safe? XTOS also has a real thread context, IIRC it's used for EDF tasks. Are we sure that we don't have code in SOF, that maps to an EDF task in XTOS and uses a mutex for locking? Or would it be safe to take a spinlock and potentially go to sleep in such a task?

Reply:
do we need to maintain XTOS ? or switch to zephyr completely ?  if so we can do nothing for k_mutex for XTOS

Reply:
@RanderWang Now I get the problem. Maybe define "struct k_mutex" so that there's both struct k_spinlock and k_spinlock_key_t as members of it when you build for XTOS. So then you can implement k_mutex_lock() with k_spin_lock() and store the key in XTOS version of "struct k_mutex". Would this work?

Reply:
@kv2019i @lgirdwood  thanks very much !  completely address my concern.

Reply:
We're already doing this in the coherent API - storing the spinlock and the key in the same persistent structure. But that made me rethink that a bit - is that actually safe, because normally you have those keys as local variables - you store your returned spinlock status (CPU interrupt state) and then you restore it from the same variable, making sure nobody else overwrites that variable in the meantime. So having that variable as persistent and available to all might sound risky - you don't want to have another context overwrite that key and then restore a wrong context. But I think it should be safe in this context because another context will only overwrite the key **after** it has taken the spinlock, so the spinlock protects the key too. Just something to be aware of. Alternatively we could wrap `k_mutex_lock()` to return a status, but I don't think we want to do that, we want to use Zephyr APIs directly.